include/turtle/detail/mock_impl.hpp

```diff
@@ -0,0 +1,108 @@
+    mutable mock::detail::function<MOCK_FUNCTION_TYPE(S)> t##_mock_;                                              \
+    mock::detail::function<MOCK_FUNCTION_TYPE(S)>& t##_mock(const mock::detail::context&,                         \
+                                                            const boost::unit_test::const_string& instance) const \
+    {                                                                                                             \
+        mock::detail::configure(*this,                                                                            \
+                                t##_mock_,                                                                        \
+                                instance.substr(0, instance.rfind(BOOST_PP_STRINGIZE(t))),                        \
+                                mock::detail::make_type_name(*this),                                              \
+                                BOOST_PP_STRINGIZE(t));                                                           \
+        return t##_mock_;                                                                                         \
+    }
+
+#define MOCK_PARAM(S) mock::detail::parameter_t < MOCK_FUNCTION_TYPE(S)
+#define MOCK_DECL_PARAM(z, n, d) BOOST_PP_COMMA_IF(n) d, n > p##n
+#define MOCK_DECL_PARAMS(n, S) BOOST_PP_REPEAT(n, MOCK_DECL_PARAM, MOCK_PARAM(S))
+#define MOCK_DECL(M, n, S, c) mock::detail::result_type_t<MOCK_FUNCTION_TYPE(S)> M(MOCK_DECL_PARAMS(n, S)) c
+
+#define MOCK_FORWARD_PARAM(z, n, d) BOOST_PP_COMMA_IF(n) d, n >> (p##n)
+#define MOCK_FORWARD_PARAMS(n, S) BOOST_PP_REPEAT(n, MOCK_FORWARD_PARAM, std::forward < MOCK_PARAM(S))
+#define MOCK_METHOD_AUX(M, n, S, t, c)                                                                    \
+    MOCK_DECL(M, n, S, c)                                                                                 \
+    {                                                                                                     \
+        static_assert(n == mock::detail::function_arity<MOCK_FUNCTION_TYPE(S)>::value, "Arity mismatch"); \
+        return MOCK_ANONYMOUS_HELPER(t)(MOCK_FORWARD_PARAMS(n, S));                                       \
+    }
+
+#define MOCK_METHOD_EXT(M, n, S, t)    \
+    MOCK_METHOD_AUX(M, n, S, t, )      \
+    MOCK_METHOD_AUX(M, n, S, t, const) \
+    MOCK_METHOD_HELPER(S, t)
+#define MOCK_CONST_METHOD_EXT(M, n, S, t) \
+    MOCK_METHOD_AUX(M, n, S, t, const)    \
+    MOCK_METHOD_HELPER(S, t)
+#define MOCK_NON_CONST_METHOD_EXT(M, n, S, t) \
+    MOCK_METHOD_AUX(M, n, S, t, )             \
+    MOCK_METHOD_HELPER(S, t)
+
+#define MOCK_FUNCTION_HELPER(S, t, s)                                                                  \
+    s mock::detail::function<MOCK_FUNCTION_TYPE(S)>& t##_mock(mock::detail::context& context,          \
+                                                              boost::unit_test::const_string instance) \
```

Comment:
is const_string something that's cheap to pass in a parameter? or is this going to allocate and copy a string?

Reply:
Yep, it is more or less `string_view`, i.e. it only contains pointers to begin and end of a string and can be convert to from any string